### ISSUE TITLE
VPNClient DNS resolver

### ISSUE THREAD

**Issue by zero88** [zero88/qwe-vpn#7]
**Final solution 20210415**
- Need `dnsmasq`
- Adapt current local DNS resolver that compatible with `dnsmasq`

================================================================
Need to check with some popular DNS resolver software to ensure:
- VPN client can ask correct DNS server 
- or provide generic solution to force VPN client query record from multiple configuration DNS server. currently it will be stopped after first DNS server record
 For example: 
   ```
   pi@n0000002:~ $ cat /etc/resolv.conf
   # Generated by resolvconf
  nameserver 192.168.15.1
  nameserver 8.8.8.8
  nameserver 169.254.169.254
  ```

Some software:
- [ ] Network manager
- [x] Connmanctl for `BBB`
- [x] [dnsmasq](https://thekelleys.org.uk/dnsmasq/doc.html)
- [x] [systemd-resolved.service](https://www.freedesktop.org/software/systemd/man/systemd-resolved.service.html)
- [x] [resolvconf](https://manpages.ubuntu.com/manpages/trusty/man8/resolvconf.8.html) in [loragateway](https://github.com/NubeIO/lorawan-complete)

Reference: 
- https://blogs.gnome.org/mcatanzaro/2020/12/17/understanding-systemd-resolved-split-dns-and-vpn-configuration/
- https://systemd.io/RESOLVED-VPNS/
- https://linux.die.net/man/8/dhclient-script
- https://man7.org/linux/man-pages/man5/resolv.conf.5.html

================== DNS Resolver ========================

Linux does not support the use of domain-specific DNS servers via `resolv.conf`. This could be potentially work around by running a caching nameserver (such as `BIND` or `dnsmasq`) locally and then configuring explicit forwarders for each domain. BBB has been bundled with `dnsmasq` and it's quite easy and straightforward to configure so we can choose it to deploy along with `vpnclient`

**Comment by zero88** [comment]
For handling dns resolver `systemd-resolved` (on ubuntu 18.04 and ubuntu 20.04), below steps could help:

**`dnsmasq` should be installed**
## Connecting to VPN network

1. Get current list of nameserver in `/run/systemd/resolve/resolv.conf`
2. Append above list to `dnsmasq` config file, below is my proposed template for `/etc/dnsmasq.d/nubeio`
```
interface=lo
bind-interfaces
listen-address=127.0.0.1
server=/$HUB_NAME.device/169.254.169.254
all-servers
```

One server address per line, following `server=` directive

3. Restart `dnsmasq` service
```
systemctl restart dnsmasq.service
```
**Now connect to VPN and get IP configuration from DHCP**

4. Get the simlink information and store it somewhere we know.
For example, store information to a text file `/app/original_resolv.conf`
```
$ readlink -f /etc/resolv.conf | sudo tee /app/original_resolv.conf
```
5. Remove the simlink and create new file with the new content: `nameserver 127.0.0.1`
```
$ sudo rm /etc/resolv.conf
$ echo "nameserver 127.0.0.1" | sudo tee /etc/resolv.conf
```

## Disconnecting from VPN network

After disconnected VPN or stopped `vpnclient` service, some cleanup tasks should be done:

1. Remove `/etc/resolv.conf` file and restore the simlink with saved information above
```
$ sudo rm /etc/resolv.conf
$ sudo ln -s $(cat /app/original_resolv.conf) /etc/resolv.conf
```
2. Restart `systemd-resolved` service to restore its original configuration
2. Stop `dnsmasq` service
3. *optional* Remove `/etc/dnsmasq.d/nubeio`
4. Run `dhclient` to refresh IP configuration (require sudo)

**Comment by zero88** [comment]
For handling `NetworkManager` (default on fedora):
1. Create below config files:
```
$ cat /etc/NetworkManager/conf.d/01-nubeio.conf
[main]
dns=dnsmasq
```
```
$ cat /etc/NetworkManager/dnsmasq.d/01-nubeio.conf
server=/$HUB_NAME.device/169.254.169.254
```
2. Verify if `dnsmasq` is preinstalled, if not install it.
3. Verify and make sure if `/etc/resolv.conf` sim link is pointing to `/run/NetworkManager/resolv.conf`
